### PR TITLE
[BugFix] array compare cast error

### PR DESCRIPTION
**Pull request by Seaven**
## Why I'm doing:
```
CREATE TABLE `t2` (
  `pk` bigint(20) NOT NULL COMMENT "",
  `aas_1` array<array<array<varchar(65533)>>> NULL COMMENT "",
  `aad_1` array<array<array<DECIMAL128(26,2)>>> NULL COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`pk`)
DISTRIBUTED BY HASH(`pk`) BUCKETS 3 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"fast_schema_evolution" = "true",
"compression" = "LZ4"
);

insert into t2 values
(1, [[["10"],["20"],["30"]],[["60"],["5"],["4"]],[["-100","-2"],["-20","10"],["100","23"]]], [[[1.00],[2.00],[3.00]],[[6.00],[5.00],[4.00]],[[-1.00,-2.00],[-2.00,10.00],[100.00,23.00]]]);

```

In centos
```
MySQL td>    select aad_1 != aas_1  from t2;
+----------------+
| aad_1 != aas_1 |
+----------------+
| 0              |
+----------------+
1 row in set
Time: 0.018s
MySQL td>
```

In ubuntu
```
MySQL td>    select aad_1 != aas_1  from t2;
+----------------+
| aad_1 != aas_1 |
+----------------+
| 1              |
+----------------+
1 row in set
Time: 0.018s
MySQL td>
```

because, memcmp result different (int -512 centos, int -2 ubuntu), and cast int32_t to int8_t error

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by stdpain — DISMISSED** [review]

File: be/src/column/array_column.cpp [hunk 1]
```diff
@@ -438,7 +438,8 @@ void ArrayColumn::compare_column(const Column& rhs_column, std::vector<int8_t>*
     size_t rows = size();
     output->resize(rows);
     for (size_t i = 0; i < rows; i++) {
-        (*output)[i] = static_cast<int8_t>(compare_at(i, i, rhs_column, 1));
+        int res = compare_at(i, i, rhs_column, 1);
+        (*output)[i] = res == 0 ? 0 : (res < 0 ? -1 : 1);
```
Comment (stdpain) [inline be/src/column/array_column.cpp:442]: res <=> 0

Comment (stdpain) [inline be/src/column/array_column.cpp:442]:
```
(res > 0) - (res < 0)
```

File: be/src/column/array_column.cpp [hunk 2]
```diff
@@ -438,7 +438,8 @@ void ArrayColumn::compare_column(const Column& rhs_column, std::vector<int8_t>*
     size_t rows = size();
     output->resize(rows);
     for (size_t i = 0; i < rows; i++) {
-        (*output)[i] = static_cast<int8_t>(compare_at(i, i, rhs_column, 1));
+        int res = compare_at(i, i, rhs_column, 1);
+        (*output)[i] = (res > 0) - (res < 0);
```
Comment (fzhedu) [inline be/src/column/array_column.cpp:442]: why not change the res from `memcmp `, but here?

File: be/src/util/memcmp.h [hunk 3]
```diff
@@ -154,7 +154,7 @@ inline int memcompare(const char* p1, size_t size1, const char* p2, size_t size2
     size_t min_size = std::min(size1, size2);
     auto res = memcmp(p1, p2, min_size);
     if (res != 0) {
-        return res;
+        return res > 0 ? 1 : -1;
```
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/util/memcmp.h:157]: I think unnecessary ? `(res > 0) - (res < 0)` is more complex

**Review by fzhedu — APPROVED** [review]

File: be/src/column/array_column.cpp (hunk 2 above)
Comment (imay) [reply be/src/column/array_column.cpp:442]: Can you add a comment here to explain why this code is written like this.

**Review by dirtysalt — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Pull request merged**